### PR TITLE
afuser: support time-reopen() 

### PR DESCRIPTION
**Pull request by alltilla**
Support `time-reopen()` option. `syslog-ng` will wait `time-reopen()` seconds (instead of the previously hard-coded 600), when it could not connect to the usertty terminal, then try to connect again.

Add per driver `time-reopen()` option, for drivers, which were previously supporting it, but was not able to set it per driver.

Signed-off-by: Attila Szakacs <attila.szakacs@oneidentity.com>

### CODE REVIEW

**Comment by kira-syslogng** [comment]
Build SUCCESS

File: modules/afuser/afuser.c
Comment (szemere) [inline modules/afuser/afuser.c:195]: Small typo here. Otherwise LGTM!

File: modules/afuser/afuser.c
Comment (Kokan) [inline modules/afuser/afuser.c:158]: questions:
1. why the default 600 is changed ?
2. if you plan to override the `time-reopen()` from global if configured, please allow local settings as well, so user could change per afuser destination.

Comment (alltilla) [reply modules/afuser/afuser.c:158]: 1. There are some problems with `time-reopen()` in general. When `GlobalConfig` is created its `time_reopen` is set to 60, and every driver will get that value. So actually it does not matter what value I put here, it will always come from the `GlobalConfig`.
2. There are other drivers, which support `time-reopen()`, but none of them support it on the driver level, only from the global `options{}` block. So I decided to not support it here, either.

So in general, I tried to follow the current `time-reopen()` logic. But your questions are valid, and I think that we should rework `time-reopen()` a bit:
1. Set `GlobalConfig`'s `time_reopen` to -1, by default.
2. Follow up this change in each driver; if they were using the `GlobalConfig` default, set their default to 60.
3. Add `KW_TIME_REOPEN` to driver level options, where the driver supports it.
4. Run tests / add new tests to check regression.

It will be a bit bigger patch, but I think this rework was due for some time. What do you think?

File: modules/afuser/afuser.c
Comment (alltilla) [reply modules/afuser/afuser.c:195]: Thanks for the review, I addressed the problems with the default value in another comment: https://github.com/syslog-ng/syslog-ng/pull/3585#discussion_r582613561

**Comment by alltilla** [comment]
I'm expecting KIRA to fail, because this probably needs some internal changes. Please do not merge, until we have prepared the internal changes.

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by alltilla** [comment]
Well this is awkward, because KIRA did not fail.. I will check if I am wrong, or we are missing some `YFLAGS` in the internal build.

File: lib/logthrdest/logthrdestdrv.c
Comment (Kokan) [inline lib/logthrdest/logthrdestdrv.c:1072]: Do not remove that check. You may know that is the case, still the compiler does not.

The logic is the following the `cfg` might be `NULL` or not. I have no clue, nor can I find out in compile time so I assume both. The next step is, well you dereference a pointer that could be `NULL`, that is undefined behavior. At that point compiler are free to do what its heart desires, and they do if they smell opportunity of performance gain. (likely this does what you would expect, which is worse because let you think it is okay).

**Review by Kokan — CHANGES_REQUESTED** [review]

File: lib/logthrdest/logthrdestdrv.c
Comment (alltilla) [reply lib/logthrdest/logthrdestdrv.c:1072]: This is new to me.

I am just thinking, why isn't this a problem in every one of our variables in every one of our functions? Like the `self` here, or we can find a lot of examples. We do not check for every variable to not be `NULL` in the beginning of our functions.

Why is this case different, and how can someone detect it?

**Comment by kira-syslogng** [comment]
Build FAILURE

File: lib/logthrdest/logthrdestdrv.c
Comment (alltilla) [reply lib/logthrdest/logthrdestdrv.c:1072]: We discussed this IRL, and added an assert to the getter.

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Review by Kokan — APPROVED** [review]

**Comment by alltilla** [comment]
Internal MR !317 is prepared. This PR is now ready to review and merge.

**Comment by amdrsantos** [comment]
@alltilla Do you know when this fix will be released? Which release and date?

**Comment by alltilla** [comment]
@amdrsantos

We will release 3.32.1 approximately 2 months from now, if everything goes well. This code will most probably be shipped with that.

From where do you obtain your syslog-ng package? Is it okay for you to use an unreleased package? As you can see it from the "Binary packages" CI run, we generate packages for every PR, and anyone can do that by themselves also with the help of `dbld`.

If you need an installer sooner, I would suggest to wait until this PR gets its 2nd approve, and gets merged, then you can generate a package from the master branch.

File: lib/logthrdest/logthrdestdrv.c
Comment (bazsi) [reply lib/logthrdest/logthrdestdrv.c:1072]: btw, I think the compiler behavior is the other way around, e.g. if it does see an unconditional dereference, it would assume the pointer is NOT null and thus drop the NULL checks, as explained here: https://lwn.net/Articles/342330/

I don't see how the code with the original change would cause suprising code to be generated.

**Review by bazsi — APPROVED** [review]
There's one note I was thinking about, namely the fact that we are peeking into LogWriterOptions in unrelated code, with expressions like "owner->writer_options->time_reopen". With that we implement an assumption that a driver is actually using LogWriter to implement its functionality.

With this we couple with an implementation detail in unrelated code. 

This does not necessarily increase coupling today, as we might have the same assumption in other parts of the same code. Also, it might be a safe assumption, as I can't see why we would want to change this implementation in the future.

I just wanted to mention it. Approve on my part, just consider the above. Thanks.

**Comment by amdrsantos** [comment]
@alltilla Thanks, I will next available release

**Comment by amdrsantos** [comment]
Hey @alltilla, I already saw that syslog-ng 3.32.1 was released.
Any idea when it will be available for Debian10 at http://download.opensuse.org/repositories/home:/laszlo_budai:/syslog-ng/Debian_10/ ?
Thanks in advance.
Alex

**Comment by alltilla** [comment]
Hi @amdrsantos,

It will probably be done in 1-2 weeks, we will send an email to the syslog-ng mail list, when it is ready.

Attila

**Comment by amdrsantos** [comment]
@alltilla do you know if Debian 10 (buster) packaging of syslog-ng 3.32.1 will be released shortly?

**Comment by alltilla** [comment]
Hi @amdrsantos 

You might have seen on the mail list, that the OBS packages are discontinued. I would advise to use `dbld` to generate the packages manually.

I generated one for you with the following commands, but you can do it manually, too:
```
wget https://github.com/syslog-ng/syslog-ng/releases/download/syslog-ng-3.32.1/syslog-ng-3.32.1.tar.gz
tar -xzvf syslog-ng-3.32.1.tar.gz
cd syslog-ng-3.32.1
./dbld/rules package-debian-buster
cd dbld/build/
zip syslog-ng-3.32.1-debian-buster-packages.zip debian-buster/*
```

We will work on providing packages in the future somewhere, but we have not decided the details yet.

[syslog-ng-3.32.1-debian-buster-packages.zip](https://github.com/syslog-ng/syslog-ng/files/6574979/syslog-ng-3.32.1-debian-buster-packages.zip)

Cheers,
Attila

**Comment by amdrsantos** [comment]
@alltilla Is it possible to use dbld to generate the packages for arm64?
If yes, do you know how?
Thanks in advance,
Alex

**Comment by bazsi** [comment]
If you are running on arm64 it should work out of the box. Cross compiling
should also be possible but is not supported atm.

**Pull request merged**
